### ISSUE TITLE
Introduce sample_weights inside fit function

### ISSUE THREAD

**Issue by filippozacchei** [dynamicslab/pysindy#645]
Context
In the general case, the training dataset may consist of trajectories with varying levels of noise or importance. In such situations, it is beneficial to weight the trajectories appropriately using a rescaling strategy (see [Weighted least squares](https://en.wikipedia.org/wiki/Weighted_least_squares)).

Proposed Solution
I would like to introduce a weighted least squares approach into the fit method, activated when sample_weights are provided.

After assembling the Θ (theta) matrix, each row can be multiplied by the square root of the corresponding sample_weight for that trajectory. This rescaling enables a straightforward implementation of a Weighted STLS algorithm.

Benefits
	•	Allows the model to account for heterogeneous noise or importance levels in the training data.
	•	Provides more robust and interpretable results in scenarios where data quality varies significantly.

**Comment by Jacob-Stevens-Haas** [comment]
This is a good ask!  If there was a place for it, I think it would be the `SampleConcatter()` object, but I also think we're moving away from a `Pipeline` based approach.  A few questions:

* weight trajectories before or after column normalization?  I'd assume before.
* I assume `score()` as well should take `sample weight`?
* Do you know if standard metadata routing, i.e. the inherited `set_fit_request()` is sufficient to allow `sample_weight` to be routed correctly through `RandomizedSearchCV` and the like?

**Comment by filippozacchei** [comment]
Thanks for the thoughtful questions! Here’s what I’m planning to do:

- Sample weights will be threaded through the pipeline so they end up in the optimizer right after `SampleConcatter` has flattened the trajectories. The weights are applied before any optional column normalization, which matches how `BaseOptimizer.fit` currently rescales `Theta` and `x_dot`.
- Yes, `score()` may have weights. I would add an explicit `sample_weight` argument, so weighted scores work identically to weighted fits.
- Metadata routing: I am not sure about this, but i [read](https://scikit-learn.org/stable/metadata_routing.html): With the Metadata Routing API, we can transfer metadata to estimators, scorers, and CV splitters using [meta-estimators](https://scikit-learn.org/stable/glossary.html#term-meta-estimators) (such as [Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline) or [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV)) or functions such as [cross_validate](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate) which route data to other objects. In order to pass metadata to a method like fit or score, the object consuming the metadata, must request it. This is done via `set_{method}_request()` methods, where {method} is substituted by the name of the method that requests the metadata. For instance, estimators that use the metadata in their `fit()` method would use `set_fit_request()`, and scorers would use `set_score_request()`. These methods allow us to specify which metadata to request, for instance `set_fit_request(sample_weight=True)`. So I suppose: once we add a sample_weight keyword to `SINDy.fit `and call `self.set_fit_request(sample_weight=True)`, meta-estimators such as `Pipeline`, `GridSearchCV`, or `RandomizedSearchCV` will pass the weights through automatically when routing is enabled. 

Let me know if anything else would help.

**Comment by Jacob-Stevens-Haas** [comment]
Much appreciated.  I think the way to navigate the metadata routing is just to have good tests to verify that the routing is happening behind the scenes.  E.g. imagine a `GridsearchCV` of `SINDy` fitting three trajectories, one with very different dynamics.  With/without `sample_weights` it should find different equations.

Are you interested in writing a PR on this?
